Neos.ContentRepository.Core/Classes/EventStore/EventPersister.php

```diff
@@ -40,16 +45,40 @@ public function publishEvents(ContentRepository $contentRepository, EventsToPubl
         $normalizedEvents = Events::fromArray(
             $eventsToPublish->events->map($this->eventNormalizer->normalize(...))
         );
-        $this->eventStore->commit(
+        $commitResult = $this->eventStore->commit(
             $eventsToPublish->streamName,
             $normalizedEvents,
             $eventsToPublish->expectedVersion
         );
+        $hooks = $this->hooksFactory->build($contentRepository);
+        $lock = $this->lockFactory->createLock($contentRepository->id->value . '_catchup');
+        $lock->acquire(true);
+
+        $expectedCheckpoint = SequenceNumber::fromInteger($commitResult->highestCommittedSequenceNumber->value - $eventsToPublish->events->count());
+
+        $projectionsToUpdate = [];
         foreach ($this->projections as $projection) {
+            if (!$projection->getCheckpoint()->equals($expectedCheckpoint)) {
+                //throw new \RuntimeException(sprintf('Projection %s is at checkpoint %d, but was expected to be at %d', $projection::class, $projection->getCheckpoint()->value, $expectedCheckpoint->value), 1714062281);
+                continue;
+            }
+            $projectionsToUpdate[] = $projection;
             if ($projection instanceof WithMarkStaleInterface) {
                 $projection->markStale();
             }
         }
-        $contentRepository->catchUpProjections();
+
+        $hooks->dispatchBeforeCatchUp();
+        foreach ($this->eventStore->load($eventsToPublish->streamName)->withMinimumSequenceNumber($expectedCheckpoint->next()) as $eventEnvelope) {
```

Comment:
This might read more events than were published..
I think we also need to specify the max sequence number!?

Reply:
No, I think it's correct this way. But it means that a newly published event might haven been applied in a separate process already